File: training/project_euler/p003_highest_factor/p003_highest_factor.py
# https://projecteuler.net/problem=3

import math

def is_prime(n):
    i = int(math.floor(math.sqrt(n)))
    prime = True
    while i > 1 and prime:
        if n % i == 0:
            prime = False
        i -= 1
    return prime

def highest_factor(n):
    i = int(math.floor(math.sqrt(n)))
    factor = n
    while i > 1 and  factor == n:
        if n % i == 0 and is_prime(i):
            factor = i
        i -= 1
    return factor

def main():
    print highest_factor(600851475143)

if __name__ == "__main__":
    main()

File: training/project_euler/p003_highest_factor/p003_highest_factor.go
// https://projecteuler.net/problem=3

package main

import "math"

func is_prime(n int) bool {
	i := int(math.Floor(math.Sqrt(float64(n))))
	prime := true
	for i > 1 && prime {
		if n%i == 0 {
			prime = false
		}
		i -= 1
	}
	return prime
}

func highest_factor(n int) int {
	i := int(math.Floor(math.Sqrt(float64(n))))
	factor := n
	for i > 1 && factor == n {
		if n%i == 0 && is_prime(i) {
			factor = i
		}
		i -= 1
	}
	return factor
}

func main() {
	print(highest_factor(600851475143))
}
